### PR TITLE
[feature/type-definitions] feat: first pass at adding type definitions

### LINKED ISSUE #56: Any plans for typescript bindings?

### PR DESCRIPTION
**Pull request by chimericdream**
closes #56 

As far as I can tell, this seems to be a relatively complete and correct type definitions file for the project. I verified it against the example app, and it appeared to be alright.

### CODE REVIEW

File: index.d.ts
Comment (kmjennison) [inline index.d.ts:59]: I'm inclined to explicitly add the required `name` field to call out that its behavior differs from the cookies library (it's a base name for multiple cookies). Open to skipping this if you don't think it's useful.

**Review by kmjennison — CHANGES_REQUESTED** [review]

File: index.d.ts
Comment (chimericdream) [reply index.d.ts:54]: Is the entire client init config optional as well? Or is the api key required for calls made from the client side no matter what?

File: index.d.ts
Comment (chimericdream) [reply index.d.ts:59]: I started using this library less than two days ago, so I will defer to you on that. I'll add it. :-)

File: index.d.ts
Comment (kmjennison) [reply index.d.ts:54]: The API key is always required. Backend calls to get refresh/ID tokens rely on it.

Comment (chimericdream) [reply index.d.ts:54]: That's what I suspected, thanks.

**Comment by chimericdream** [comment]
I just noticed that the return type of `withAuthUserTokenSSR` (and possibly a couple of the others) isn't quite right. Don't merge this yet.

**Comment by chimericdream** [comment]
I'm going to push one more small change, exporting the `GetSSRProps` utility type. That will simplify adding type info to the prop getter you can pass into the `with...SSR` functions.

**Comment by chimericdream** [comment]
It might be worth waiting a day or so to merge this. I'm finding quite a few little things that need tidying up. I think it's pretty close, though.

**Comment by Qs-F** [comment]
IMO, it would be better to use string `union` type instead of `enum`.  ref: http://g.page/typescript+don%27t+use+enums
As written in TypeScript handbook,
> Enums are one of the few features TypeScript has which is not a type-level extension of JavaScript.

ref: https://www.typescriptlang.org/docs/handbook/enums.html
enums are located in the middle of TypeScript and JavaScript, and this is little bit troublesome.

So here is my suggestion:

```ts
const AuthAction = {
  RENDER: "render",
  SHOW_LOADER: "showLoader", 
  RETURN_NULL: "returnNull",
  REDIRECT_TO_LOGIN: "redirectToLogin",
  REDIRECT_TO_APP: "redirectToApp",
} as const;

export type AuthAction = typeof AuthAction[keyof typeof AuthAction];
```

**Comment by chimericdream** [comment]
@Qs-F that's a good suggestion, thanks.

**Comment by chimericdream** [comment]
@kmjennison I have ironed out all the wrinkles I could find. If you want to leave this open a little longer for more feedback, that's fine with me. But for my part I think it's ready to merge whenever you like.

**Comment by Qs-F** [comment]
@chimericdream
Thank you!
I'm so sorry, but I don't know the code does work inside `*.d.ts` …
I see, that makes sense.

Sorry for interrupting!

**Comment by chimericdream** [comment]
Nothing to be sorry for. I didn't realize it wouldn't work either.

**Review by kmjennison — APPROVED** [review]

**Pull request merged**
